### PR TITLE
[byon] Implement backend

### PR DESCRIPTION
**Pull request by tumido**
Since we're hurrying with this, I've implemented it all at once (commits have slight dependency, so it goes as a single PR) 

Resolves https://github.com/opendatahub-io/odh-dashboard/issues/153
Resolves https://github.com/opendatahub-io/odh-dashboard/issues/142
Resolves https://github.com/opendatahub-io/odh-dashboard/issues/152
Resolves https://github.com/opendatahub-io/odh-dashboard/issues/151
Resolves https://github.com/opendatahub-io/odh-dashboard/issues/143

### CODE REVIEW

**Comment by tumido** [comment]
This PR also updates the API - adds fields for `id` and `software`.

File: backend/src/types.ts [hunk 1]
```diff
@@ -228,18 +234,78 @@ export type NotebookError = {
...
+export type NotebookCreateRequest = {
+  name: string;
+  url: string;
+  description?: string;
+}
+
+export type NotebookUpdateRequest = {
+  id: string;
   visible?: boolean;
   packages?: NotebookPackage[];
```
Comment (dlabaj) [inline backend/src/types.ts:253]: Can this be deleted if we have software?

File: backend/src/types.ts [hunk 2]
```diff
@@ -31,17 +31,23 @@ export declare type QuickStart = {
...
+
+export type K8sResourceCommon = {
   metadata?: {
     name?: string;
     namespace?: string;
+    generateName?: string;
     uid?: string;
     labels?: { [key: string]: string };
     annotations?: { [key: string]: string };
+    creationTimestamp?: Date;
```
Comment (dlabaj) [inline backend/src/types.ts:47]: Thinking maybe should be a number instead.  The time in miliseconds.

File: backend/src/types.ts [hunk 3]
```diff
@@ -228,18 +234,78 @@ export type NotebookError = {
 export type NotebookStatus = "Importing" | "Validating" | "Succeeded" | "Failed";
 
 export type Notebook = {
-  name: string;
-  url: string;
-  description?: string;
+  id: string;
   phase?: NotebookStatus;
   user?: string;
   uploaded?: Date;
```
Comment (dlabaj) [inline backend/src/types.ts:240]: number instead of date?

**Review by dlabaj — COMMENTED** [review]
looks good a few comments.

**Review by dlabaj — APPROVED** [review]

File: backend/src/types.ts (hunk 1 above)
Comment (tumido) [reply backend/src/types.ts:253]: I don't think so. There are 2 distinct content annotatations available in JH per ImageStream, see: 

https://github.com/opendatahub-io/odh-manifests/blob/9ddfec0e5950203acb17d8560378b3c9d67de57e/jupyterhub/notebook-images/overlays/additional/elyra-notebook-imagestream.yaml#L16-L17

`opendatahub.io/notebook-software` maps to the `software` field here and `opendatahub.io/notebook-python-dependencies` maps to `packages. See also https://odh-io.slack.com/archives/C037Q2HLDED/p1648114485074139?thread_ts=1647632733.034829&cid=C037Q2HLDED

**Pull request merged**